wreck: ensure cancelled jobs are no longer in active job hash

grondo opened the pull request:
This is a partial fix of #1504, once flux-framework/flux-sched#334 is closed then the fix will be complete.

I had to relax the requirement for `kvspath` in the wreck.state payload in `wreck_state_cb`, since the scheduler will not easily have access to the kvs path at the time the job is cancelled. This should not be a problem since kvs_path should not be used when processing the terminal state of a job anyway.

grondo commented:
I should probably add a test here, but to do so would require the dummy sched module growing a `sched.cancel` interface... Hm...

garlick commented on src/modules/wreck/job.c:
```diff
@@ -758,7 +758,7 @@ static void wreck_state_cb (flux_t *h, flux_msg_handler_t *w,
     struct wreck_job *job;
 
-    if (flux_event_unpack (msg, &topic, "{s:I s:s}",
+    if (flux_event_unpack (msg, &topic, "{s:I s?s}",
```
initialize kvs_path to NULL in case it's not assigned?

grondo replied:
Ah, good catch!

garlick commented:
> I should probably add a test here, but to do so would require the dummy sched module growing a sched.cancel interface... Hm...

That seems straightforward.  There is also this:

Can `flux wreck cancel --force` be used as a test?
```
Usage: flux-wreck cancel [-f|--force] JOBID
Cancel a pending job
Options:
  -h, --help                 Display this message.
  -f, --force                Force cancel even if scheduler is not loaded
```

Maybe in addition to sending the `wreck,<jobid>.kill` event, it could send a status update event?

grondo commented:
> Can flux wreck cancel --force be used as a test?

No, the help message is misleading. `-f, --force` just falls back to `flux wreck kill` if `sched.cancel` returns ENOSYS.

grondo commented:
> Maybe in addition to sending the wreck,<jobid>.kill event, it could send a status update event?

That might be racy, since the `kill` propagates as a signal to tasks which then are reaped by wrexecd, which finally sets the job to "complete". The "cancelled" state can only be reached from "submitted" state (for this implementation anyway)

grondo commented:
Rebased on current master, resolved @garlick's review comments, and added a test in `t2000-wreck-dummy-sched` to ensure cancelled jobs are removed from the active jobs hash. (Used an event generated by hand for testing)

garlick commented:
Restarted a builder that was stuck in `t4000-issues-test-driver.t`.  I'll merge when that completes.

garlick commented:
Hmm, the gcc4.9 builder has failed twice in a row here:
```
[0;32mPASS[m: t2000-wreck-dummy-sched.t 1 - load dummy sched module
[0;32mPASS[m: t2000-wreck-dummy-sched.t 2 - job.sumbit issues correct event
[0;32mPASS[m: t2000-wreck-dummy-sched.t 3 - job.submit-nocreate issues correct event
[0;35mERROR[m: t2000-wreck-dummy-sched.t - missing test plan
[0;35mERROR[m: t2000-wreck-dummy-sched.t - exited with status 1
```
I can't seem to find anything in the output that supplies clues.  Will restart it again.

grondo commented:
Ok, duh, it was a chain-lint error in that new test. Sorry!

Rebased on current master and fixed the error.

grondo commented:
This time hit `not ok 53 - instance can stop cleanly with subscribers (#1025)`. Restarted.

garlick commented:
Green!  Yay!
